be2-scala/src/main/scala/ch/epfl/pop/pubsub/graph/validators/ElectionValidator.scala

```diff
@@ -470,4 +535,131 @@ sealed class ElectionValidator(dbActorRef: => AskableActorRef) extends MessageDa
+    }
+  }
+
+  /** @param electionQuestionResult
+    *   The ElectionQuestionResult the ElectionValidator received.
+    * @return
+    *   The total number number of votes on that question.
+    */
+  private def countNumberOfVotes(electionQuestionResult: ElectionQuestionResult): Int = {
+    electionQuestionResult.result.map(_.count).sum
+  }
+
+  /** @param electionQuestionResult
+    *   The ElectionQuestionResult the ElectionValidator received.
+    * @return
+    *   True if all the ballot counts are greater that 0, false otherwise.
+    */
+  private def allBallotCountsArePositive(electionQuestionResult: ElectionQuestionResult): Boolean = {
+    electionQuestionResult.result.forall(electionBallotVotes => electionBallotVotes.count >= 0)
+  }
+
+  /** checks that the question ids received with the ResultElection are coherent with the ones received in the SetupElection
+    * @param rpcMessage
+    *   the rpc message to validate.
+    * @param questions
+    *   The list of ElectionQuestions that have been sent with the SetupElection.
+    * @param result
+    *   The list of ElectionQuestionResults the ElectionValidator received.
+    * @param electionId
+    *   The electionId to compute the election question ids.
+    * @param error
+    *   The error to forward in case where ballot options are incoherent.
+    * @return
+    *   GraphMessage: passes the rpcMessages to Right if successful else Left with pipeline error.
+    */
+  private def checkResultQuestionIds(
+      rpcMessage: JsonRpcRequest,
+      questions: List[ElectionQuestion],
+      result: List[ElectionQuestionResult],
+      electionId: Hash,
+      error: PipelineError
+  ): GraphMessage = {
+    val setupElectionIds = questions.map(_.id).toSet
+    val resultElectionIds = result.map(_.id).toSet
+    if (setupElectionIds == resultElectionIds) {
+      Right(rpcMessage)
+    } else
+      Left(error)
+  }
+
+  /** checks that for the given ElectionResult, the ballot options match the ballot options sent in the SetupElection message.
+    * @param rpcMessage
+    *   The rpc message to validate.
+    * @param questions
+    *   The list of ElectionQuestions that have been sent with the SetupElection.
+    * @param result
+    *   The list of ElectionQuestionResults the ElectionValidator received.
+    * @param error
+    *   The error to forward in case where ballot options are incoherent.
+    * @return
+    *   GraphMessage: passes the rpcMessages to Right if successful else Left with pipeline error.
+    */
+  private def checkResultBallotOptions(
+      rpcMessage: JsonRpcRequest,
+      questions: List[ElectionQuestion],
+      result: List[ElectionQuestionResult],
+      error: PipelineError
+  ): GraphMessage = {
+    var isResultBallotValid = true
+    result.foreach(electionQuestionResult => {
```

Comment:
Instead of using `foreach`, you should use `forall` since you want that the condition inside the loop is true for all questions. 
Your code will be more concise ;)